### PR TITLE
fix: round to correct precision when step uses exponential notation

### PR DESCRIPTION
**Pull request by lucasweng**
Closes https://github.com/adobe/react-spectrum/issues/8274

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 📝 Test Instructions:

- Added tests for `roundToStepPrecision`.
- Existing tests should still pass.

### CODE REVIEW

File: packages/@react-stately/utils/test/number.test.ts [hunk 1]
```diff
@@ -8,6 +8,30 @@ describe('number utils', () => {
     });
   });
 
+  describe('roundToStepPrecision', () => {
+    it('should return the input unchanged for integer steps', () => {
+      expect(roundToStepPrecision(7.123, 1)).toBe(7.123);
+      expect(roundToStepPrecision(5, 10)).toBe(5);
+    });
+    it('should round to the correct decimal places for steps with decimals', () => {
+      expect(roundToStepPrecision(1.24, 0.1)).toBe(1.24);
```
Comment (snowystinger) [inline packages/@react-stately/utils/test/number.test.ts:17]: maybe I'm over thinking this, but shouldn't this be
```suggestion
      expect(roundToStepPrecision(1.24, 0.1)).toBe(1.2);
```

**Comment by scrgl** [comment]
Hi @lucasweng, thank you for taking the time to look into this!

I believe it would be helpful to add a few more test cases to cover edge scenarios. Specifically, since JavaScript automatically converts very small numbers into exponential notation, it would be great to include a case like 0.0000000231 as a step value. This gets transformed into "2.31e-8", and your current implementation may not handle it correctly—it identifies the dot and treats the value as if it were in standard decimal format.

To simplify things and make the logic more robust, I’d suggest normalizing all numbers to exponential notation upfront. That way, you only need to handle one consistent format.

Here's a snippet that might help—though please double-check it against your requirements and test cases:
```
function getPrecision(num: number): number {
    const e = num.toExponential();
    const [base, exponent] = e.split('e');
    const [, decimalPart] = base.split('.');
    return decimalPart
      ? decimalPart.length - parseInt(exponent)
      : Math.abs(parseInt(exponent));
  }
```

File: packages/@react-stately/utils/test/number.test.ts (hunk 1 above)
Comment (lucasweng) [reply packages/@react-stately/utils/test/number.test.ts:17]: hi @scrgl thanks for pointing out that incorrect implementation! https://github.com/adobe/react-spectrum/pull/8290/commits/872c52702bae55fa43008b23976ae0d0503e6f41 fixed the issue mentioned in https://github.com/adobe/react-spectrum/pull/8290#issuecomment-2929973595 and added more tests.

I opted not to use `split` as suggested in the snippet because it creates additional intermediate arrays every time `snapValueToStep` is called. Since this function may be invoked multiple times by hooks and components, those arrays could quickly add up. Also, since `step` values without explicit exponential notation are likely more common in user land, normalizing everything to exponential format upfront would add extra overhead for those cases.

File: packages/@react-stately/utils/src/number.ts [hunk 2]
```diff
@@ -20,9 +20,22 @@ export function clamp(value: number, min: number = -Infinity, max: number = Infi
...
-  let precision = pointIndex >= 0 ? stepString.length - pointIndex : 0;
+  // Handle negative exponents in exponential notation (e.g., "1e-7" → precision 8)
+  let eIndex = stepString.toLowerCase().indexOf('e-');
+  if (eIndex > 0) {
+    let significand = stepString.slice(0, eIndex);
+    let exponent = stepString.slice(eIndex + 1);
+    let pointIndex = significand.indexOf('.');
+    let decimals = pointIndex >= 0 ? significand.length - pointIndex : 0;
+    precision = Math.abs(Number(exponent)) + decimals + 1;
+  } else {
```
Comment (snowystinger) [inline packages/@react-stately/utils/src/number.ts]: I think this is a little more succint and explains where the `+ 1` was coming from
```suggestion
  if (eIndex > 0) {
    precision = Math.abs(Math.floor(Math.log10(Math.abs(step)))) + eIndex;
  } else {
```

Comment (snowystinger) [reply packages/@react-stately/utils/src/number.ts]: though don't know if it's susceptible to floating point precision issues... i haven't thought of a case yet

at the least, we can replace the `+ decimals + 1` with eIndex

Comment (lucasweng) [reply packages/@react-stately/utils/src/number.ts]: https://github.com/adobe/react-spectrum/pull/8290/commits/c447134dbbb3cd7830af01559b6fa95e1f1f7359 Thanks 🙏🏻

**Review by snowystinger — APPROVED** [review]
